### PR TITLE
Fix brittle assertion error in cache tests (resolves #911)

### LINKED ISSUE #911: Fix brittle and obscure assertion in jobCacheTest
See my comments on PR #898.

https://github.com/BD2KGenomics/toil/pull/898#discussion-diff-65220948

### PR DESCRIPTION
**Pull request by arkal**
resolves #911

Setup new error and catch that in the tests instead

### CODE REVIEW

File: src/toil/job.py [hunk 1]
```diff
@@ -1288,9 +1288,10 @@ def deleteLocalFile(self, fileStoreID):
                                     logger.debug(fileToDelete,  'was read mutably and deleted by '
                                                  'the user')
                                 else:
-                                    raise CacheError('Illegal operation detected. Cache tracked'
-                                                     'file deleted explicitly by user. Use'
-                                                     'deleteLocalFile to delete such files.')
+                                    raise CacheIllegalDeleteError('Cache tracked file deleted '
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Starting the string on the next line should reduce the number of wraps within the string.

Comment (hannes-ucsc) [inline src/toil/job.py]: Actually the message should be defined in the constructor, not the call to the constructor.

Comment (hannes-ucsc) [inline src/toil/job.py]: You might want to include the offending file name in the message. But I am not sure.

File: src/toil/job.py [hunk 2]
```diff
@@ -2454,6 +2455,14 @@ def __init__(self, message):
         super(CacheError, self).__init__(message)
 
 
+class CacheIllegalDeleteError(Exception):
+    '''
+    Error Raised if the Toil detects the user deletes a cached file
+    '''
+
+    def __init__(self, message):
+        super(CacheIllegalDeleteError, self).__init__(message)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: This constructor currently does nothing. You should define the message here.

File: src/toil/job.py [hunk 3]
```diff
@@ -2454,6 +2455,14 @@ def __init__(self, message):
         super(CacheError, self).__init__(message)
 
 
+class CacheIllegalDeleteError(Exception):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Rename to IllegalDeletionCacheError

File: src/toil/job.py [hunk 4]
```diff
@@ -2454,6 +2455,14 @@ def __init__(self, message):
...
+    '''
+    Error Raised if the Toil detects the user deletes a cached file
+    '''
+
+    def __init__(self, message):
+        super(CacheIllegalDeleteError, self).__init__(message)
+
 class CacheInvalidSrcError(Exception):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Rename InvalidSourceCacheError and make subclass of CacheError.

File: src/toil/job.py (hunk 3 above)
Comment (hannes-ucsc) [inline src/toil/job.py]: ... and make subclass of CacheError.

File: src/toil/test/src/jobCacheTest.py [hunk 5]
```diff
@@ -334,7 +335,7 @@ def _parseAssertionError(logFile):
             Parse the assertion error message from a failed toil logfile
 
             :param logFile: path to the logfile
-            :return: Str of the error message
+            :return: tuple of Error type, Str of the error message
```
Comment (hannes-ucsc) [inline src/toil/test/src/jobCacheTest.py]: What is up with the crazy capitalization and abbrevation? Use English!

File: src/toil/test/src/jobCacheTest.py [hunk 6]
```diff
@@ -244,11 +244,12 @@ def _testCacheEviction(self, file1MB, file2MB, diskRequestMB):
...
                 self.assertEqual(err.numberOfFailedJobs, 1)
-                errMsg = self._parseAssertionError(self.options.logFile)
-                if errMsg == 'Unable to free up enough space for caching.':
+                errType, errMsg = self._parseAssertionError(self.options.logFile)
+                if (errType == 'AssertionError' and
+                        errMsg == 'Unable to free up enough space for caching.'):
                     self.assertEqual(expectedResult, 'Fail')
                 else:
-                    self.fail('Shouldn\'t see this')
+                    self.fail('An assertionError was not thrown by Toil')
```
Comment (hannes-ucsc) [inline src/toil/test/src/jobCacheTest.py]: Why is assertionError lower-case?

File: src/toil/test/src/jobCacheTest.py [hunk 7]
```diff
@@ -704,15 +705,16 @@ def testRemoveLocalImmutablyReadFile(self):
...
             try:
                 Job.Runner.startToil(A, self.options)
             except FailedJobsException as err:
                 self.assertEqual(err.numberOfFailedJobs, 2)
-                errMsg = self._parseAssertionError(self.options.logFile)
-                if 'explicitly' not in errMsg:
-                    self.fail('Shouldn\'t see this')
+                errType, errMsg = self._parseAssertionError(self.options.logFile)
+                if errType != 'CacheIllegalDeleteError':
+                    self.fail('Illegal delete operation not picked up by test.')
```
Comment (hannes-ucsc) [inline src/toil/test/src/jobCacheTest.py]: use self.asserEquals()

File: src/toil/job.py (hunk 1 above)
Comment (arkal) [inline src/toil/job.py]: Good point with the file name. I'll add that in

File: src/toil/job.py (hunk 3 above)
Comment (arkal) [inline src/toil/job.py]: right-o

**Comment by arkal** [comment]
@hannes-ucsc The failing jobs are unrelated.

**Pull request merged**